#+SETUPFILE: orgsetup.org

* =ensembldb=: coordinate mapping between the genome, exome, transcriptome and proteome

*EuroBioc2018*, December 2018, Munich

Johannes Rainer (Eurac Research, Italy)
johannes.rainer@eurac.edu
github/twitter: /jotsetung/

** =ensembldb= and =EnsDb= databases

- =ensembldb=: create and use Ensembl-based annotation resources.
- Genomic *and* protein annotations available.
- Powerful filter framework.
- *Update*: functions to map between coordinate systems:
  - =genomeToTranscript=
  - =transcriptToGenome=
  - =transcriptToProtein=
  - =proteinToTranscript=
  - =proteinToGenome=
  - =genomeToProtein=
    
** Example: map genomic to protein coordinates

- Variant at position 16:89920138 related to human red hair color.
- _Task_: find the amino acid that is affected by this variant.
  #+BEGIN_SRC R
    library(ensembldb)
    library(EnsDb.Hsapiens.v86)
    library(magrittr)

    ## Change chromosome naming style to UCSC
    seqlevelsStyle(EnsDb.Hsapiens.v86) <- "UCSC"

    ## Map genomic to protein coordinates
    prt_pos <- GRanges("chr16", IRanges(89920138, width = 1)) %>%
        genomeToProtein(EnsDb.Hsapiens.v86)
    prt_pos

    ## Get the amino acid sequences for the proteins
    prt_seq <- EnsDb.Hsapiens.v86 %>%
        filter(~ protein_id == names(prt_pos[[1]])) %>%
        proteins(return.type = "AAStringSet")
    prt_seq

    ## Get the amino acid at position 294
    library(Biostrings)
    subseq(prt_seq, start = start(prt_pos)[[1]], width = 1)

    ## Answer: affected amino acid: aspartic acid
  #+END_SRC

** Example: map protein-relative positions to the genome

- Background
  - Down syndrome characterized by a third copy of chromosome 21.
  - Phenotype most likely caused by gene-dosage effect.
  - _Task_: find genes with a specific protein domain, plot transcript(s)
    highlighting the exons encoding the protein domains.

- Search proteins with /helix-loop-helix/ domain on chromosome 21.
  #+BEGIN_SRC R
    ## Get genes on chromosome 21 encoding a HLH domain
    gns <- EnsDb.Hsapiens.v86 %>%
        filter(~ seq_name == "chr21" & protein_domain_id == "PF00010") %>%
        genes()
    gns

    ## Get transcript models for SIM2 encoding PF00010
    txs <- EnsDb.Hsapiens.v86 %>%
        filter(~ gene_name == "SIM2" &
                   protein_domain_id == "PF00010") %>%
        getGeneRegionTrackForGviz
    txs
  #+END_SRC

- Fetch all Pfam protein domains for the /SIM2/ transcript.
  #+BEGIN_SRC R
    ## Fetch all Pfam protein domains for the transcript(s)
    pdoms <- EnsDb.Hsapiens.v86 %>%
        filter(~ tx_id %in% txs$transcript &
                      protein_domain_source == "pfam") %>%
        proteins(columns = c("protein_domain_id", "prot_dom_start",
                             "prot_dom_end"))
    pdoms
  #+END_SRC

- Map protein domain coordinates to the genome.
  #+BEGIN_SRC R
    pdoms_rng <- IRanges(start = pdoms$prot_dom_start,
                         end = pdoms$prot_dom_end,
                         names = pdoms$protein_id)

    ## Map protein domains to genomic coordinates
    pdoms_gnm <- proteinToGenome(pdoms_rng, EnsDb.Hsapiens.v86)
    pdoms_gnm
  #+END_SRC

- Plot transcript models and encoded protein domains.
  #+BEGIN_SRC R
    ## Convert the list to a GRanges with grouping information
    pdoms_grng <- unlist(GRangesList(pdoms_gnm))
    pdoms_grng$id <- rep(pdoms$protein_domain_id, lengths(pdoms_gnm))
    pdoms_grng$grp <- rep(1:nrow(pdoms), lengths(pdoms_gnm))
    pdoms_grng



    library(Gviz)
    ## Define the individual tracks:
    ## - Genome axis
    gaxis_trk <- GenomeAxisTrack()
    ## - Transcripts
    gene_trk <- GeneRegionTrack(txs, showId = TRUE, just.group = "right",
                                name = "", geneSymbol = TRUE, size = 0.5)
    ## - Protein domain track
    pdom_trk <- AnnotationTrack(pdoms_grng, group = pdoms_grng$grp,
                                id = pdoms_grng$id, groupAnnotation = "id",
                                just.group = "right", shape = "box",
                                name = "Protein domains", size = 0.5)

    plotTracks(list(gaxis_trk, gene_trk, pdom_trk))
  #+END_SRC

** Finally

*Thank you for your attention!*

Problems? Questions? -> https://github.com/jotsetung/ensembldb

